### PR TITLE
[ARTS-601] Validation of Site Service using Hierarchy Rules

### PR DESCRIPTION
**Pull request by kedarl-rdb**
## Description
This PR has got scenarios covered for validation for creating New Trial Site API 

### Changes
- [ARTS-182] - create New Trial Site API 

### Checklist:

- [X] Branch name follows convention (feature/arts-#-short-name OR fix/arts-#-short-name)
- [X] I have included a short-meaningful title, description and changes for this PR

[ARTS-182]: https://ndph-arts.atlassian.net/browse/ARTS-182

### CODE REVIEW

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/SiteServiceApplication.java [hunk 1]
```diff
@@ -13,6 +13,10 @@
     
     private static final Logger LOGGER = LoggerFactory.getLogger(SiteServiceApplication.class);
 
+    /**
```
Comment (alexbaxter) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/SiteServiceApplication.java:16]: Not sure this comment adds anything

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/converter/OrganizationConverter.java [hunk 2]
```diff
@@ -30,6 +30,9 @@
             fhirOrganization.setId(new IdType(input.getSiteId()));
         }
         setParentOrganization(input, fhirOrganization);
+        if (input.getSiteType() != null) {
+            fhirOrganization.setImplicitRules(input.getSiteType());
```
Comment (alexbaxter) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/converter/OrganizationConverter.java:34]: Has this been tested against the FHIR server? The implictRules attribute has to be a valid URI - this is fairly loose, but I believe has to start with a scheme identifier (e.g. we could use sitetype:CCO)

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/model/Attribute.java [hunk 3]
```diff
@@ -48,6 +50,10 @@ public static Attribute fromString(String input) throws InitialisationError {
             return NAME;
         } else if (AttributeNames.ALIAS.nameof().equals(input)) {
             return ALIAS;
+        } else if (AttributeNames.PARENT_SITE_ID.nameof().equals(input)) {
```
Comment (alexbaxter) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/model/Attribute.java:53]: I don't like this else-if tower, can we just use some method on the AttributeNames enum instead?

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/model/SiteConfiguration.java [hunk 4]
```diff
@@ -31,9 +31,30 @@ public SiteConfiguration(String name, String displayName, List<SiteAttributeConf
...
+     * Site Constructor with two parameters
+     *
+     * @param name the Site name
+     * @param displayName the Site displayName
+     * @param type the Site type
+     * @param attributes the List of SiteAttributeConfiguration attributes
+     * @param child the List of SiteAttributeConfiguration child / children.
+     *
+     */
+    public SiteConfiguration(String name, String displayName, String type,
```
Comment (alexbaxter) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/model/SiteConfiguration.java:44]: There are now 3 constructors for SiteConfiguration, can we just have one? In real use, only the default constructor is used I think (for deserialization) - the other constructors are just used by tests (which could use the setters instead).

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/repository/FhirRepository.java [hunk 5]
```diff
@@ -42,18 +42,16 @@
      * @return all organization instances in the store, might be empty, not null
      */
     List<Organization> findOrganizations();
-
+
```
Comment (alexbaxter) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/repository/FhirRepository.java:45]: This just seems like a spurious commit, no substantive changes - suggest reverting it.

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java [hunk 6]
```diff
@@ -25,17 +29,22 @@
...
+     * @param configurationProvider provide Site Configuration
      * @param siteStore        Site store interface
      * @param entityValidation Site validation interface
      */
     @Autowired
-    public SiteServiceImpl(final EntityStore<Site, String> siteStore,
+    public SiteServiceImpl(SiteConfigurationProvider configurationProvider,
+                           final EntityStore<Site, String> siteStore,
                            ModelEntityValidation<Site> entityValidation) {
+        var configuration = configurationProvider.getConfiguration();
```
Comment (alexbaxter) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:47]: Why no null check on configurationProvider and configurationProvider.getConfiguration()?

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java [hunk 7]
```diff
@@ -44,22 +53,40 @@ public SiteServiceImpl(final EntityStore<Site, String> siteStore,
         }
         this.siteStore = siteStore;
         this.entityValidation = entityValidation;
+        initMaps(configuration);
         logger.info(Services.STARTUP.message());
     }
 
+    private void initMaps(final SiteConfiguration configuration) {
```
Comment (alexbaxter) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:60]: Don't need initMaps(), all the work is done in the recursive addTypesToMap method which can be invoked directly from the constructor.

**Review by alexbaxter — COMMENTED** [review]
Not approving or rejecting now because I think there are some more commits to be done on this PR.

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/SiteServiceApplication.java (hunk 1 above)
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/SiteServiceApplication.java:16]: Fixes the check style Error

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java (hunk 7 above)
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:60]: Agree, removed the initMaps()

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/converter/OrganizationConverter.java (hunk 2 above)
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/converter/OrganizationConverter.java:34]: I have tested this against FHIR and it seems to hold

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/model/SiteConfiguration.java (hunk 4 above)
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/model/SiteConfiguration.java:30]: Removed the unused constructor

**Review by kedarl-rdb — COMMENTED** [review]
Addressed few Review Comments

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java (hunk 6 above)
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:46]: Will add the null chceck

**Review by alexbaxter — DISMISSED** [review]
OK to merge now

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java [hunk 8]
```diff
@@ -25,49 +28,78 @@
...
+        var configuration = configurationProvider.getConfiguration();
         this.siteStore = siteStore;
         this.entityValidation = entityValidation;
+        addTypesToMap(configuration);
         logger.info(Services.STARTUP.message());
     }
 
+    private void addTypesToMap(final SiteConfiguration configuration) {
+        this.sitesByType.put(configuration.getType(), configuration);
+        if (configuration.getChild() != null) {
```
Comment (darrengsmithson) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:64]: could use CollectionsUtils library or similar to do null check to make code tidier..e.g emptyIfNull(collection).foreach

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java [hunk 9]
```diff
@@ -25,49 +28,78 @@
...
+        }
+    }
+
     /**
      * @param site the Site to save.
      * @return The id of the new site
      */
     @Override
     public String save(final Site site) {
         var validationResponse = entityValidation.validate(site);
```
Comment (darrengsmithson) [inline site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:78]: couldnt we just call a validate method and teh validate method throws the exceptions? I think the 'if' checks should be in the validator?

Comment (alexbaxter) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:78]: We agreed that the validator should validate just the entity by itself, and the service does the checks on parent-child relationships etc. - that way the validator doesn't depend on the entity store.

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java [hunk 10]
```diff
@@ -25,49 +28,78 @@
 @Service
 public class SiteServiceImpl implements SiteService {
 
+    private final Map<String, SiteConfiguration> sitesByType = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (alexbaxter) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:31]: Should be one value per site type (key) - this is the lookup for site types.

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java (hunk 9 above)
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:78]: Agree with Alex

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/model/Attribute.java [hunk 11]
```diff
@@ -11,7 +11,9 @@
 public enum Attribute {
 
     NAME(AttributeNames.NAME.nameof(), Site::getName),
```
[the remark this answers is not in the document]
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/model/Attribute.java:13]: Myself and Alex neither
As suggested renamed the field to value

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/Services.java [hunk 12]
```diff
@@ -10,7 +10,9 @@
     SITE_NOT_FOUND("Site ID not found"),
     PARENT_NOT_FOUND("Parent site ID not found"),
     SITE_EXISTS("Site with given name exists"),
-    ONE_ROOT_SITE("Root site already exists");
+    ONE_ROOT_SITE("Root site already exists"),
```
[the remark this answers is not in the document]
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/Services.java:13]: As suggested renamed the same

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java [hunk 13]
```diff
@@ -25,49 +28,78 @@
...
     private final Logger logger = LoggerFactory.getLogger(SiteServiceImpl.class);
 
     /**
+     * @param configurationProvider provide Site Configuration
      * @param siteStore        Site store interface
      * @param entityValidation Site validation interface
      */
     @Autowired
-    public SiteServiceImpl(final EntityStore<Site, String> siteStore,
+    public SiteServiceImpl(SiteConfigurationProvider configurationProvider,
```
[the remark this answers is not in the document]
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:43]: Agree

File: site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java (hunk 8 above)
Comment (kedarl-rdb) [reply site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java:64]: Used CollectionsUtils.emptyIfNull(collection)

**Review by alexbaxter — APPROVED** [review]
OK to merge

**Review by darrengsmithson — APPROVED** [review]

**Pull request merged**